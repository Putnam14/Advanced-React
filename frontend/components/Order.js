import { Component } from 'react';
import { Query } from 'react-apollo';
import { format } from 'date-fns';
import Head from 'next/head';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { SINGLE_ORDER_QUERY } from '../queries/queries';
import formatMoney from '../lib/formatMoney';
import Dump from './Dump';
import Error from './ErrorMessage';

const OrderStyles = styled.div`
  max-width: 1000px;
  margin: 0 auto;
  border: 1px solid ${props => props.theme.offWhite};
  box-shadow: ${props => props.theme.bs};
  padding: 2rem;
  border-top: 10px solid red;
  & > p {
    display: grid;
    grid-template-columns: 1fr 5fr;
    margin: 0;
    border-bottom: 1px solid ${props => props.theme.offWhite};
    span {
      padding: 1rem;
      &:first-child {
        font-weight: 900;
        text-align: right;
      }
    }
  }
  .order-item {
    border-bottom: 1px solid ${props => props.theme.offWhite};
    display: grid;
    grid-template-columns: 300px 1fr;
    align-items: center;
    grid-gap: 2rem;
    margin: 2rem 0;
    padding-bottom: 2rem;
    img {
      width: 100%;
      height: 100%;
      object-fit: cover;
    }
  }
`;

class Order extends Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
  };

  render() {
    return (
      <Query
        query={SINGLE_ORDER_QUERY}
        variables={{ id: this.props.id }}
        fetchPolicy="network-only"
      >
        {({ data, error, loading, refetch }) => {
          if (loading) return <p>Loading...</p>;
          if (error) return <Error error={error} refetch={refetch} />;
          const order = data.order;
          return (
            <OrderStyles data-test="order">
              <Head>
                <title>Sick Fits - Order {order.id}</title>
              </Head>
              <p>
                <span>Order Id:</span>
                <span>{order.id}</span>
              </p>
              <p>
                <span>Charge</span>
                <span>{order.charge}</span>
              </p>
              <p>
                <span>Date</span>
                <span>{format(order.createdAt, 'MMMM D, YYYY h:mm A')}</span>
              </p>
              <p>
                <span>Order Total</span>
                <span>{formatMoney(order.total)}</span>
              </p>
              <p>
                <span>Item Count</span>
                <span>{order.items.length}</span>
              </p>
              <div className="items">
                {order.items.map(item => (
                  <div className="order-item" key={item.id}>
                    <img src={item.image} alt={item.title} />
                    <div className="item-details">
                      <h2> {item.title} </h2>
                      <p>Qty: {item.quantity}</p>
                      <p>Each: {formatMoney(item.price)}</p>
                      <p>Subtotal: {formatMoney(item.price * item.quantity)}</p>
                      <p>SubTotal: {item.description}</p>
                      <p>{item.description}</p>
                    </div>
                  </div>
                ))}
              </div>
            </OrderStyles>
          );
        }}
      </Query>
    );
  }
}

export default Order;
