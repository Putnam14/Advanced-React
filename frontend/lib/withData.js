import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { LOCAL_STATE_QUERY } from '../queries/queries.graphql';
// can also be a function that accepts a `headers` object (SSR only) and returns a config

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? 'http://localhost:4444' : 'http://localhost:4444',
    ssrMode: !process.browser, // Disables forceFetch on the server (so queries are only run once)
    request: async operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    clientState: {
      resolvers: {
        Mutation: {
          toggleCart(_, variables, { cache }) {
            const { cartOpen } = cache.read({ query: LOCAL_STATE_QUERY });
            const data = { data: { cartOpen: !cartOpen } };
            cache.writeData(data);
            return data;
          },
        },
      },
      defaults: {
        cartOpen: false,
      },
    },
  });
}

export default withApollo({ client: createClient });
