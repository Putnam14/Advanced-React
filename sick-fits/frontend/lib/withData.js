// Next-Apollo package to make sure server-side rendering works
import withApollo from 'next-with-apollo'
// You can import Apollo along with any 'link's you want, or do this
import ApolloClient from 'apollo-boost'
import { endpoint } from '../config'

function createClient({ headers }) {
  return new ApolloClient({
    // Can change this if you have different Yoga APIs for dev, prod
    uri: process.env.NODE_ENV === 'development' ? endpoint : endpoint,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          // Bring cookies along for the ride
          credentials: 'include',
        },
        headers,
      })
    },
  })
}

// Use to create client in _app.js
export default withApollo(createClient)
