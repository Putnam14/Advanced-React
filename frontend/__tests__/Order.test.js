import React from 'react';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from 'react-apollo/test-utils';
import Order, { SINGLE_ORDER_QUERY } from '../components/Order';
import { fakeOrder } from '../lib/testUtils';

const mocks = [
  {
    request: { query: SINGLE_ORDER_QUERY, variables: { id: '123' } },
    result: {
      data: {
        order: fakeOrder(),
      },
    },
  },
];

describe('<Order/>', () => {
  it('Renders loading state', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Order id="123" />
      </MockedProvider>
    );
    expect(wrapper.text()).toContain('Loading...');
  });

  it('Renders the order', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Order id="123" />
      </MockedProvider>
    );
    await wait(50);
    wrapper.update();
    const order = wrapper.find('div[data-test="order"]');
    expect(toJSON(order)).toMatchSnapshot();
  });
});
