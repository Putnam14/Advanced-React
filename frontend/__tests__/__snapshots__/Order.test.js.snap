// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Order/> Renders with Proper Data 1`] = `
Array [
  <styled.div
    data-test="order"
  >
    <div
      className="sc-bwzfXH wJijb"
      data-test="order"
    >
      <SideEffect(Head)>
        <Head />
      </SideEffect(Head)>
      <p>
        <span>
          Order Id:
        </span>
        <span>
          3a430a73-c9e9-4bcc-b46a-41618965ffea
        </span>
      </p>
      <p>
        <span>
          Charge
        </span>
        <span>
          Hello World
        </span>
      </p>
      <p>
        <span>
          Date
        </span>
        <span>
          May 13, 2014 12:53 PM
        </span>
      </p>
      <p>
        <span>
          Order Total
        </span>
        <span>
          $24.04
        </span>
      </p>
      <p>
        <span>
          Item Count
        </span>
        <span>
          2
        </span>
      </p>
      <div
        className="items"
      >
        <div
          className="order-item"
          key="f17558f5-7760-4116-acd2-76da805ab010"
        >
          <img
            alt="Ut assumenda sint"
            src="dog.jpg"
          />
          <div
            className="item-details"
          >
            <h2>
               
              Ut assumenda sint
               
            </h2>
            <p>
              Qty: 
              8
            </p>
            <p>
              Each: 
              $500
            </p>
            <p>
              Subtotal: 
              $4,000
            </p>
            <p>
              SubTotal: 
              Exercitationem omnis laborum exercitationem est sapiente laudantium. Assumenda quasi adipisci mollitia iste itaque omnis.
            </p>
            <p>
              Exercitationem omnis laborum exercitationem est sapiente laudantium. Assumenda quasi adipisci mollitia iste itaque omnis.
            </p>
          </div>
        </div>
        <div
          className="order-item"
          key="f05cef40-52ca-4dca-b480-7618d01ced82"
        >
          <img
            alt="Similique temporibus"
            src="dog.jpg"
          />
          <div
            className="item-details"
          >
            <h2>
               
              Similique temporibus
               
            </h2>
            <p>
              Qty: 
              9
            </p>
            <p>
              Each: 
              $500
            </p>
            <p>
              Subtotal: 
              $4,500
            </p>
            <p>
              SubTotal: 
              Sapiente soluta corporis commodi. Nisi cum ducimus. Doloremque maiores esse atque et.
            </p>
            <p>
              Sapiente soluta corporis commodi. Nisi cum ducimus. Doloremque maiores esse atque et.
            </p>
          </div>
        </div>
      </div>
    </div>
  </styled.div>,
  <div
    className="sc-bwzfXH wJijb"
    data-test="order"
  >
    <SideEffect(Head)>
      <Head />
    </SideEffect(Head)>
    <p>
      <span>
        Order Id:
      </span>
      <span>
        3a430a73-c9e9-4bcc-b46a-41618965ffea
      </span>
    </p>
    <p>
      <span>
        Charge
      </span>
      <span>
        Hello World
      </span>
    </p>
    <p>
      <span>
        Date
      </span>
      <span>
        May 13, 2014 12:53 PM
      </span>
    </p>
    <p>
      <span>
        Order Total
      </span>
      <span>
        $24.04
      </span>
    </p>
    <p>
      <span>
        Item Count
      </span>
      <span>
        2
      </span>
    </p>
    <div
      className="items"
    >
      <div
        className="order-item"
        key="f17558f5-7760-4116-acd2-76da805ab010"
      >
        <img
          alt="Ut assumenda sint"
          src="dog.jpg"
        />
        <div
          className="item-details"
        >
          <h2>
             
            Ut assumenda sint
             
          </h2>
          <p>
            Qty: 
            8
          </p>
          <p>
            Each: 
            $500
          </p>
          <p>
            Subtotal: 
            $4,000
          </p>
          <p>
            SubTotal: 
            Exercitationem omnis laborum exercitationem est sapiente laudantium. Assumenda quasi adipisci mollitia iste itaque omnis.
          </p>
          <p>
            Exercitationem omnis laborum exercitationem est sapiente laudantium. Assumenda quasi adipisci mollitia iste itaque omnis.
          </p>
        </div>
      </div>
      <div
        className="order-item"
        key="f05cef40-52ca-4dca-b480-7618d01ced82"
      >
        <img
          alt="Similique temporibus"
          src="dog.jpg"
        />
        <div
          className="item-details"
        >
          <h2>
             
            Similique temporibus
             
          </h2>
          <p>
            Qty: 
            9
          </p>
          <p>
            Each: 
            $500
          </p>
          <p>
            Subtotal: 
            $4,500
          </p>
          <p>
            SubTotal: 
            Sapiente soluta corporis commodi. Nisi cum ducimus. Doloremque maiores esse atque et.
          </p>
          <p>
            Sapiente soluta corporis commodi. Nisi cum ducimus. Doloremque maiores esse atque et.
          </p>
        </div>
      </div>
    </div>
  </div>,
]
`;
